File: PAT乙级/数字黑洞.cpp
#include <bits/stdc++.h>

using namespace std;

int main(int argc, char **argv){
    int i, a, b;
    string s;
    vector<char> v1, v2;
    cin >> i;
    while(1){
        s = to_string(i);
        v1.assign(s.begin(),s.end());
        sort(v1.begin(), v1.end());  //从小到大排序
        v2 = v1;
        reverse(v1.begin(), v1.end());

        s = ""; for( auto i : v1) s += i;  a = stoi(s);
        s = ""; for( auto i : v2) s += i;  b = stoi(s);
        i = a - b;
        for(auto i: v1) cout << i ;
        cout << " - ";
        for (auto i: v2) cout << i ;
        cout << " = " << i << endl; 
        if (i == 6174) break;
    }
    return 0;
}

File: PAT乙级/数字黑洞.py
i = input()
i += '0'*(4 - len(i))
if(len(set(list(i))) == 1):
    print(i + " - " + i + " = 0000")
    exit()
while(i != 6174):
    L = list(str(i))
    
    L.sort( reverse = True)
    a = int("".join(L))
    L1 = L[::-1]
    b = int("".join(L1))
    i = a - b
    print("".join(L) +" - " + "".join(L1) + " =", i)

